arbeitszeit/use_cases/list_registered_hours_worked.py

```diff
@@ -0,0 +1,54 @@
+from dataclasses import dataclass
+from datetime import datetime
+from decimal import Decimal
+from uuid import UUID
+
+from arbeitszeit.records import AccountTypes
+from arbeitszeit.repositories import DatabaseGateway
+
+
+@dataclass
+class Request:
+    company_id: UUID
+
+
+@dataclass
+class RegisteredHoursWorked:
+    hours: Decimal
+    worker_id: UUID
+    worker_name: str
+    registered_on: datetime
+
+
+@dataclass
+class Response:
+    registered_hours_worked: list[RegisteredHoursWorked]
+
+
+@dataclass
+class ListRegisteredHoursWorkedUseCase:
+    database_gateway: DatabaseGateway
+
+    def list_registered_hours_worked(self, request: Request) -> Response:
+        company = (
+            self.database_gateway.get_companies().with_id(request.company_id).first()
+        )
+        assert company
+        work_account = company.get_account_by_type(AccountTypes.a)
+        assert work_account
+        records = (
+            self.database_gateway.get_transactions()
```

Comment:
Wait a second 🤔, why does the "Use Cases" (Interactor) layer contain a gateway from the outer "Interface Adapters" layer? I think the inner layers don't know the outer layers; the outer layers only operate on the inner layer via Dependency Injection (DI). Please compare with [this diagram](https://blog.cleancoder.com/uncle-bob/2012/08/13/the-clean-architecture.html) by Uncle Bob. Or am I missing something? I'm not proficient in Clean Architecture™, so my understanding of it might be wrong.

Reply:
Also, [the article](https://blog.cleancoder.com/uncle-bob/2012/08/13/the-clean-architecture.html) below the linked diagram might be of interest, e.g.:

> [...]
> The Dependency Rule
> [...]
> The overriding rule that makes this architecture work is The Dependency Rule. This rule says that source code dependencies can only point inwards. Nothing in an inner circle can know anything at all about something in an outer circle. In particular, the name of something declared in an outer circle must not be mentioned by the code in the an inner circle. That includes, functions, classes. variables, or any other named software entity.
> [...]
> Use Cases
> [...]
> We do not expect changes in this layer to affect the entities. We also do not expect this layer to be affected by changes to externalities such as the database, the UI, or any of the common frameworks. This layer is isolated from such concerns.
> [...]

Reply:
Hey, 
important point, I hope I can clarify it.

We import the `DatabaseGateway` from the "business logic" module: `from arbeitszeit.repositories import DatabaseGateway`. This DatabaseGateway is an interface (a `Protocol`). We do NOT import the SQL database implementation (`arbeitszeit_flask/database/repositories.py`) nor the in-memory implementation we use for tests (`tests/use_cases/repositories.py`). We inject these implementations when needed.

Then there is still the question: Does these protocol classes know something about SQL database implementations? Some wordings ("join with", "offset", "limit", etc.) seem to point in that direction. But actually I think this abstract interface can be implemented by any kind of repository, be it in-memory, sql, or anything else. That we need some kind of repository in the business logic should be clear.  

Does that make sense and clarify this point?

Reply:
Mhh 🤔, I think, I'm in the wrong here! I had a look, in the [Clean Architecture](https://www.oreilly.com/library/view/clean-architecture-a/9780134494272/) book. There in chapter 22 (page 202ff.) on page 208 is a structural (kind of) UML diagram (figure 22.2) which shows that the Use Case Interactor indeed can have a navigable direction towards a Data Access Interface which is comparable to our use of the Database Gateway Protocol (structural sub-typing). Also the text in that chapter seems to confirm this. I will push the state of my review to "approved", once again 😊!